### ISSUE TITLE
abstract classes are present in choices lists

### ISSUE THREAD

**Issue by skinkie** [tefra/xsdata#625]
I have noticed that abstract classes are present in choices lists. I think this is not as _intended_.

**Comment by tefra** [comment]
Any example ?

**Comment by skinkie** [comment]
> Any example ?

You will not be happy when I say NeTEx GeneralFrame members...

**Comment by tefra** [comment]
Hey man do you want to point me to a type in the schema, which exact compound field you mean and why do you think its wrong?

I like scavenger hunts but yeah netex schemas are just ridiculous :laughing:

**Comment by tefra** [comment]
> ...so I think you cannot use an abstract member in a new document, right?

ok let's assume here that's is true you still have this spaghetti of substitution groups

```xml
<xsd:element name="PricingRule" substitutionGroup="PricingRule_">
```

```xml
<xsd:element name="PricingRule_" type="DataManagedObjectStructure" abstract="true" substitutionGroup="DataManagedObject">
```


```xml
<xsd:complexType name="generalFrameMembers_RelStructure">
<xsd:annotation>
	<xsd:documentation>A collection of one or more VERSION FRAME MEMBERS.</xsd:documentation>
</xsd:annotation>
<xsd:complexContent>
	<xsd:extension base="containmentAggregationStructure">
		<xsd:choice maxOccurs="unbounded">
			<xsd:element ref="GeneralFrameMember"/>
			<xsd:element ref="DataManagedObject"/>
			<xsd:element ref="Entity_Entity"/>
		</xsd:choice>
	</xsd:extension>
</xsd:complexContent>
```

The model needs to list as a choice the `PricingRule_` even if it means any derived type there.

**Comment by skinkie** [comment]
Lets say the spaghetti originates from the model that implements inheritance. But I don't see how an abstract="true" can ever be instantiated as an XML element. So in my perspective abstract=true should not be in the list. The derivative, using the substitutionGroup, should ofcourse.

**Comment by tefra** [comment]
Yes abstract  elements can't appear in an xml document instance, but derived types can. If we remove abstract types from all the generated python typing annotations we also loose the ability to assign / parse any derived types. 


The Netex schemas use the abstract elements with the substitution groups as a way to "easily" extend various other types, If we exclude the abstract `PricingRule_` from the substitution groups then we will also miss the substitutions that depend on that like

```xml
<xsd:element name="PricingRule" substitutionGroup="PricingRule_">
<xsd:element name="DiscountingRule" substitutionGroup="PricingRule_">
<xsd:element name="LimitingRule" substitutionGroup="PricingRule_">
```

I am trying to figure out a logic here and how we could incorportate it in the generator without breaking the first rule with the derived types.

**Comment by tefra** [comment]
You are on to something @skinkie I made a test to remove all fields that are based on abstract elements and all my tests pass just fine in all my samples repo and the w3c test suite.

I need to dig into this a little bit more, I want to see if this is supposed to apply to complex types as well and I want to check how xjc handles this as well.

**Comment by skinkie** [comment]
Thanks again for following up.

**Comment by tefra** [comment]
Thanks for reporting @skinkie after running some tests I am quite confident its correct, thanks for reporting!

I know its like a drop in the ocean but for the netex collection it saved like 5k lines of code :rocket: 

closed in #627

**Comment by skinkie** [comment]
> Thanks for reporting @skinkie after running some tests I am quite confident its correct, thanks for reporting!

Thanks again!

> I know its like a drop in the ocean but for the netex collection it saved like 5k lines of code rocket

Especially for the NeTEx code some 'lazy loading' would still be something that I would like to see. For development I am using the 'clustering' style, it is better than the single-package, but still takes "time" to start the application.
